### PR TITLE
get Arch Info using HSA

### PR DESCRIPTION
**Pull request by micmelesse**
this pr uses HSA code to query arch info and warpsize instead of excepting a rocminfo installation on the machine.

### CODE REVIEW

**Comment by dfukalov** [comment]
It seems the functionality to request device properties is already implemented at https://github.com/ROCmSoftwarePlatform/triton/blob/triton-mlir/python/triton/compiler.py#L2284-L2298

So perhaps we can just add needed fields to hip_utils.get_device_properties interface...

**Review by rsanthanam-amd — APPROVED** [review]
LGTM.

File: CMakeLists.txt [hunk 1]
```diff
@@ -240,16 +240,20 @@ if(TRITON_BUILD_PYTHON_MODULE)
     MLIRIR
   )
 
+  set(ROCM_LIBRARIES
+    ${CMAKE_CURRENT_SOURCE_DIR}/lib/rocm/libhsa-runtime64.so
```
[the remark this answers is not in the document]
Comment (binarman) [reply CMakeLists.txt:244]: I mean, this line nails libtriton.so to particular library from source codes, so you can not move it anywhere.
I am not sure how to deploy this kind of change in pypi as a prebuild package, for example.

**Comment by dfukalov** [comment]
Would you please clarify what is the reason to not use libraries and headers from /opt/rocm ?

**Comment by rsanthanam-amd** [comment]
> Would you please clarify what is the reason to not use libraries and headers from /opt/rocm ?

This is a change that was requested by Phillipe to eliminate the need to rely on /opt/rocm or to define an environment variable specifying the root rocm path.

The request was to eliminate the need to rely on the `rocminfo` utility to discern arch string, arch triple, etc.

Also, Pytorch 2.0 relies on triton and our internal team has been having issues with packaging triton into an integrated pypi wheel.

This approach integrates the necessary bits directly into the triton source and eliminates the need to call `rocminfo`.

/cc @micmelesse

**Comment by dfukalov** [comment]
It seems to me (i) removing the whole dependency on /opt/rocm contents, (ii) implementing a way to configure where rocm installed and (iii) removing rocminfo dependency are three separate topics.

As for latter one (iii) I already suggested in https://github.com/ROCmSoftwarePlatform/triton/pull/181#issuecomment-1495588135 the much simpler approach just to add all/needed fields in already implemented interface.

**Comment by micmelesse** [comment]
> It seems to me (i) removing the whole dependency on /opt/rocm contents, (ii) implementing a way to configure where rocm installed and (iii) removing rocminfo dependency are three separate topics.
> 
> As for latter one (iii) I already suggested in [#181 (comment)](https://github.com/ROCmSoftwarePlatform/triton/pull/181#issuecomment-1495588135) the much simpler approach just to add all/needed fields in already implemented interface.

I will look into it.

**Comment by dfukalov** [comment]
Btw, as for (ii) I see at least in some cases `ROCM_PATH` env variable is already used. E.g. without the variable difined triton cannot find ld.lld to link .hsaco. So I guess it also may be used for headers and libraries path specifying.

**Comment by rsanthanam-amd** [comment]
> Btw, as for (ii) I see at least in some cases `ROCM_PATH` env variable is already used. E.g. without the variable difined triton cannot find ld.lld to link .hsaco. So I guess it also may be used for headers and libraries path specifying.

Yes, that is true.  our internal pytorch 2.0 team is looking into the ld.lld issue.

**Comment by micmelesse** [comment]
I looked into using hip to get the info. I can get the gfx arch and warp number but not sramecc and snack settings. The current solution might be ugly but it is complete.

File: include/triton/Tools/Sys/GetPlatform.hpp [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+      return HSA_STATUS_ERROR_OUT_OF_RESOURCES;
+    }
+
+    // fill data
+    err = hsa_isa_get_info_alt(agent_i.agent_isa, HSA_ISA_INFO_NAME,
+                               isa_i.name_str);
+
+    // set output tuple
+    std::get<0>(*arch_info) = isa_i.name_str;
+    std::get<1>(*arch_info) = agent_i.wavefront_size;
```
Comment (dfukalov) [inline include/triton/Tools/Sys/GetPlatform.hpp:284]: Actually I do not see sramecc and xnack functionality in the PR. I see only name and wavefront size.

Comment (rsanthanam-amd) [reply include/triton/Tools/Sys/GetPlatform.hpp:284]: We get this info from the gfx_arch_details.

https://github.com/ROCmSoftwarePlatform/triton/blob/ee0cf024290b512a9201977ab0aee20230708dc1/python/triton/compiler.py#L1738

Comment (binarman) [reply include/triton/Tools/Sys/GetPlatform.hpp:284]: Daniil means that this data is not stored in arch triple.
Features are flags passed to compiler, like `-march=gfx908:sramecc+:xnack-`

For example, I have `gfx1030` on my local machine, which supports following features: `cumode` and `wavefrontsize64` according to [this table](https://www.llvm.org/docs/AMDGPUUsage.html#processors), but rocminfo prints following:
```
$ rocminfo | grep amd
      Name:                    amdgcn-amd-amdhsa--gfx1030
```

Comment (rsanthanam-amd) [reply include/triton/Tools/Sys/GetPlatform.hpp:284]: there does not seem to be a way to get this information (sramecc and xnack settings) using daniil's suggested approach.

**Pull request merged**